### PR TITLE
Remove LINQ allocations for all batched sends

### PR DESCRIPTION
**Pull request by danielmarbach**
# All SDK Contribution checklist:

This checklist is used to make sure that common guidelines for a pull request are followed.
- [ ] **Please open PR in `Draft` mode if it is:**
	- Work in progress or not intended to be merged.
	- Encountering multiple pipeline failures and working on fixes.
- [ ] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md).**
- [x] **The pull request does not introduce [breaking changes](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md).**

### [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#sdk-generation-guidelines)
- [ ] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as the commitid of your swagger spec or link to the swagger spec, used to generate the code. (Track 2 only)
- [ ] The `*.csproj` and `AssemblyInfo.cs` files have been updated with the new version of the SDK. Please double check nuget.org current release version.

## Additional management plane SDK specific contribution checklist: 
Note: Only applies to `Microsoft.Azure.Management.[RP]` or `Azure.ResourceManager.[RP]`
 
- [ ] Include updated [management metadata](https://github.com/Azure/azure-sdk-for-net/tree/main/eng/mgmt/mgmtmetadata).
- [ ] Update AzureRP.props to add/remove version info to maintain up to date API versions.

### Management plane SDK Troubleshooting
- If this is very first SDK for a services and you are adding new service folders directly under /SDK, please add `new service` label and/or contact assigned reviewer.
- If the check fails at the `Verify Code Generation` step, please ensure:
	- Do not modify any code in generated folders.
	- Do not selectively include/remove generated files in the PR.
	- Do use `generate.ps1/cmd` to generate this PR instead of calling `autorest` directly.
	Please pay attention to the @microsoft.csharp version output after running `generate.ps1`. If it is lower than current released version (2.3.82), please run it again as it should pull down the latest version.
	
	**Note: We have recently updated the PSH module called by `generate.ps1` to emit additional data. This would help reduce/eliminate the Code Verification check error. Please run following command**:

	    `dotnet msbuild eng/mgmt.proj /t:Util /p:UtilityName=InstallPsModules`

### Old outstanding PR cleanup
 Please note:
	If PRs (including draft) has been out for more than 60 days and there are no responses from our query or followups, they will be closed to maintain a concise list for our reviewers.

### CODE REVIEW

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Primitives/ServiceBusModelFactory.cs
Comment (danielmarbach) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Primitives/ServiceBusModelFactory.cs:507]: This is a necessary evil since any arbitrary `IList<T>` can be passed that might not be directly castable to `IReadOnlyCollection<T>`

**Comment by JoshLove-msft** [comment]
> @JoshLove-msft @jsquire if I didn't screw things up this will be a pretty nice allocation improvement. Need to stop though for today. If you are planning to "sneak" this in too, feel free to edit the PR directly.

Thanks @danielmarbach! For this one, I think we can wait until the next release as I've already started the release pipeline.
I'd also be interested in whether these improvements also apply to Event Hubs /cc @jsquire

**Comment by jsquire** [comment]
I probably won't get a chance to dig into this until tomorrow, but at a glance it's something I can likely apply to Event Hubs as well... so double win!

**Comment by danielmarbach** [comment]
Now that we return a proper materialized list from the batches, it would also be possible to switch for example the DiagnosticExtensions away from using IEnumerable to clearly indicate there will be no multiple enumeration problem. Should I do that too?

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpMessageBatch.cs
Comment (jsquire) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpMessageBatch.cs:157]: We only used `IEnumerable` because the legacy client used it and we didn't reevaluate after mimicking the pattern.  Should we consider just returning the source type (List<T>) here and avoid any cast?   Its internal use and we have no plans to expose it wider and the callers are trusted.   On the other hand, this form does protect us from accidental mutation.    Thoughts?

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Primitives/ServiceBusModelFactory.cs
Comment (jsquire) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Primitives/ServiceBusModelFactory.cs:507]: I agree it's not impactful due to the testing context, but should we add a small bit of logic to streamline when we can?   Something like:

```csharp
public override IReadOnlyCollection<T> AsReadOnly<T>() =>
    _backingStore switch
    { 
        IReadOnlyCollection collection => collection,
        _ => new List<T>((IEnumerable<T>)_backingStore)
    };
```

**Review by jsquire — COMMENTED** [review]
LGTM.   Thanks, @danielmarbach!   I'll let Josh have the final word.

**Comment by jsquire** [comment]
> Now that we return a proper materialized list from the batches, it would also be possible to switch for example the DiagnosticExtensions away from using IEnumerable to clearly indicate there will be no multiple enumeration problem. Should I do that too?

I vote yes.  I don't see us needing to loosen that guarantee.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpMessageBatch.cs
Comment (danielmarbach) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpMessageBatch.cs:157]: At first I had to leave it with `IReadOnlyCollection` because there was an overload that passed in an array with a single message. Now that I have introduced an overload that accepts just one message and allows bypassing that additional array allocation, it would be possible to just expose the `List<T>` directly as you mentioned. 

From a performance perspective, it would be slightly better because we can make sure the right enumerator is used. Not sure how powerful the latest devirtualization improvements are and whether it matters that much. I'd personally also buy into the trusted caller perspective and go without the abstraction. I can see also the benefit of the abstraction since the data returned there is consumed by multiple internal method calls.

**Comment by danielmarbach** [comment]
Green 🥳 🎉

**Comment by danielmarbach** [comment]
Let me know about the other comments. Can push it here or do another PR. Fine for me either way

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpMessageBatch.cs
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpMessageBatch.cs:157]: Ah, I missed this discussion. Let me take a look and get back to you on this.

Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpMessageBatch.cs:157]: I think we can go ahead with the change to `IList<T>`. 

One thing I noticed (and it looks like Jesse is already handling in his EH PR) is that if we do this, we would need to still have the SendBatchInternalAsync method in AmqpSender take an IEnumerable to handle the [public API](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/servicebus/Azure.Messaging.ServiceBus/src/Sender/ServiceBusSender.cs#L193) that takes an IEnumerable events which we then convert to a readonlylist. It looks like in Jesse's PR he is adding a new overload to [AmqpMessageConverter](https://github.com/Azure/azure-sdk-for-net/pull/26983/files#diff-da252bede9fe91fffbfd42d0f1f8df3a051fa54640d936a71602535fce9b10bcR82) so that we can still benefit from the perf enhancements for the MessageBatch case.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Sender/ServiceBusSender.cs
Comment (danielmarbach) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Sender/ServiceBusSender.cs:408]: I guess one drawback the change from `IReadOnlyCollection` to `List` has is that previously if someone passed an array or a list we could directly pass it through, since both implement `IReadOnlyCollection` if I'm not mistaken. Now for all calls where the enumerable is an array we would allocate a list. Judging by that I think the original IReadOnlyCollection approach I had was smarter. Can you double-check my assumptions?

Comment (danielmarbach) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Sender/ServiceBusSender.cs]: I double-checked and indeed List and Array both provide IReadOnlyCollection. Therefore using that as the inner type seems smarter because then we don't start allocating for probably the cases when users pass an array as an enumerable. So I switched back in commit https://github.com/Azure/azure-sdk-for-net/pull/26911/commits/19cf8fc0c47c520b453b02b75be9be8cbfeaf689

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpMessageBatch.cs
Comment (jsquire) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpMessageBatch.cs:157]: Agreed; I had forgotten that we were eagerly materializing the enumerable at the call site.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpMessageConverter.cs
Comment (jsquire) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpMessageConverter.cs:99]: Can L98 just use `firstMessage`?

**Review by jsquire — APPROVED** [review]

**Pull request merged**
